### PR TITLE
Add an apiFetch middleware to automatically handle media upload failures

### LINKED ISSUE #17474: Fix handling of HTTP 500 errors while uploading images
The main reason an image upload would fail is the server running out of resources while post-processing (creating sub-sizes). In WP 5.3 the resizing of images after upload was updated/fixed and it is now possible to retry creating the sub-sizes, even after an upload request ends with a server "crash" (timeout or out-of-memory fatal error).

On the client side an unique "upload ID" is sent with each image upload request. Then if the original request ends with HTTP 500 error, the client does another request including the same upload ID so the server (PHP) can try to create any missing image sub-sizes. This can be repeated several times.

Then, if all attempts fail, the client sends a "cleanup" request so the server can delete any orphaned sub-sizes, etc. The client should also show appropriate error message, something like:

**Post-processing of the image failed. If this is a photo or a large image, please scale it down to 2500 pixels and upload it again.**

### PR DESCRIPTION
**Pull request by youknowriad**
closes #17474

Implements the new Media "post-process" endpoint to recover from media upload failures

**Testing instructions**

See here https://github.com/WordPress/gutenberg/issues/17474#issuecomment-539607050

or https://github.com/WordPress/gutenberg/pull/17858#issuecomment-540114688

> Perhaps another (simpler) way to test is to add the following to a WP plugin:
> ```
> add_filter( 'intermediate_image_sizes_advanced', '_test_image_post_processing' );
> add_filter( 'wp_get_missing_image_subsizes', '_test_image_post_processing' );
> function _test_image_post_processing( $sizes ) {
> 	if ( rand( 1, 2 ) === 2 ) {
> 		trigger_error( '', E_USER_ERROR );
> 		exit;
> 	}
> 	
> 	return $sizes;
> }
> ```
> 
> It will fail 50% of the time and should trigger multiple post_process request.

With that in place, you either drag & drop an image from your desktop or use the upload button.

### CODE REVIEW

File: packages/api-fetch/src/index.js [hunk 1]
```diff
@@ -80,48 +77,9 @@ const defaultFetchHandler = ( nextOptions ) => {
 		}
 	);
 
-	const parseResponse = ( response ) => {
```
Comment (youknowriad) [inline packages/api-fetch/src/index.js]: The parsing and error normalizing here was extracted to be reused in middlewares.

**Comment by azaozz** [comment]
Trying to test this in several ways, works properly when the initial upload request fails, but couldn't get it to do more than one `post_process` request.

The `X-WP-Upload-Attachment-ID` header is set for all uploads, but post-processing is currently only possible for images. Perhaps it would be good to check whether the upload is an image before sending a request.

Also, (assuming this is not in yet), if all (5) attempts fail, the new attachment will have to be deleted, and the error should tell the user to "scale down the image and try uploading it again".

**Comment by azaozz** [comment]
Perhaps another (simpler) way to test is to add the following to a WP plugin:
```
add_filter( 'intermediate_image_sizes_advanced', '_test_image_post_processing' );
add_filter( 'wp_get_missing_image_subsizes', '_test_image_post_processing' );
function _test_image_post_processing( $sizes ) {
	if ( rand( 1, 2 ) === 2 ) {
		trigger_error( '', E_USER_ERROR );
		exit;
	}
	
	return $sizes;
}
```

It will fail 50% of the time and should trigger multiple post_process request.

**Comment by youknowriad** [comment]
> The X-WP-Upload-Attachment-ID header is set for all uploads, but post-processing is currently only possible for images. Perhaps it would be good to check whether the upload is an image before sending a request.

I guess the request will just fail for these other media. I wonder if it's better to not check. Potentially, we might need post processing for other medias in the future. The check in a generic way in the middleware might be a bit complex to do. How are you checking in the media library?

**Comment by youknowriad** [comment]
Also, I personally feel that instead of randomly trying to post process 5 times, the server should tell us that it's possible to post process or not using a header.

**Comment by youknowriad** [comment]
Another question:  is it possible that a failure request don't contain the "ID"? Said differently, for "delete" and "post-process" requests, should we use the last id received or just keep the first one and use it for all follow-up requests?

**Comment by youknowriad** [comment]
Also potentially,  the endpoint should return the "post-process" URL as a header and the client could just "POST" to that URL without any data. It would be more generic.

**Comment by getsource** [comment]
> > The X-WP-Upload-Attachment-ID header is set for all uploads, but post-processing is currently only possible for images. Perhaps it would be good to check whether the upload is an image before sending a request.
> 
> I guess the request will just fail for these other media. I wonder if it's better to not check. Potentially, we might need post processing for other medias in the future. 

PDFs are the other one I currently know about.
WP also does processing of Audio and Video files. For example, with MP3s, ID3 data is extrated, including the creation of a thumbnail (attachment created). I suspect this round wouldn't handle separating any of that, but thought it might be useful information.

**Comment by azaozz** [comment]
(Response to https://github.com/WordPress/gutenberg/pull/17858#issuecomment-540446233)

> I guess the request will just fail for these other media. I wonder if it's better to not check. 

Yeah, that's a possibility too.

> Potentially, we might need post processing for other medias in the future.

The new end-point requires the `create-image-subsizes`  "action" which (logically) limits it only to image uploads. The idea was to add other actions for other file types that may need post-processing.

Looking there again, we can check what type of file was uploaded on the PHP side, no need to get that info from the client.

> The check in a generic way in the middleware might be a bit complex to do. How are you checking in the media library?

The Media Library uses Plupload, it checks the file type (sets mime type prop) before uploading.

**Comment by azaozz** [comment]
(Response to https://github.com/WordPress/gutenberg/pull/17858#issuecomment-540446556)

> Also, I personally feel that instead of randomly trying to post process 5 times, the server should tell us that it's possible to post process or not using a header.

Yes, it does that. The client should look in the error type/error message returned from the server. If it is an HTTP 500 error (meaning a fatal error in PHP) it can assume there is more work to be done on the server.

If the error is 400, 403, etc. there's no need to do another request. In these cases there is also a "proper" response from the API.

The random/arbitrary part here is how many times to retry to do the post-processing when there is a PHP fatal error. Typically it can fail because the server is busy at that moment, or because there's not enough memory available. In the first case it would take 30 seconds for each request. In the second case it will fail fairly quickly. However there may be other reasons for the error, and post-processing may succeed on a second/third attempt. Seems 3 to 5 times is a good rule of thumb. This can probably be enhanced further in the future.

**Comment by azaozz** [comment]
(Reply to https://github.com/WordPress/gutenberg/pull/17858#issuecomment-540467205)

> Another question: is it possible that a failure request don't contain the "ID"? Said differently, for "delete" and "post-process" requests, should we use the last id received or just keep the first one and use it for all follow-up requests?

Yes, "delete" and "post-process" requests should use the "already known" attachment ID. Earlier version of the API patch did "pass through" the ID, but later that part was removed. If that makes it easier/better lets add it back (just need to add the custom header in `post_process_item()`).

**Comment by azaozz** [comment]
(Reply to https://github.com/WordPress/gutenberg/pull/17858#issuecomment-540469306)

> Also potentially, the endpoint should return the "post-process" URL as a header and the client could just "POST" to that URL without any data. It would be more generic.

Ideally, yes. However there are some architectural difficulties on the API side to do that in time, before the expected PHP fatal errors. That may get fixed/enhanced in the future.

**Comment by azaozz** [comment]
To summarize:
1. TBD: shall we remove the  `create-image-subsizes` action from the `/post-process` end-point? Looking at it again, it's not strictly necessary to have an "action" there. The client can decide whether to retry post-processing after a HTTP 500 error, and the server can "see" what kind of file was just uploaded and take appropriate action.
 2. TBD: shall we pass-through the new attachment ID on requests to the `/post-process` end-point?

**Comment by azaozz** [comment]
(Reply to https://github.com/WordPress/gutenberg/issues/17474#issuecomment-537494119)

> Any thoughts on the implementation? Is this something that should be a wp.apiFetch middleware or implemented more specifically per use-case?

Looking at the code, given all the "requirements" mentioned above, I guess it would be nice if the middleware doesn't try to handle these errors "automatically", but instead "pass them up the stack" in some way?

It would be nice to have some more logic on the client side to determine when to do another `/post-process` request depending on the error type and on uploaded file type? Then that additional logic can be in the "upload" component which can expose it and better determine what happens.

Or perhaps this can be some form of a "callback" on all HTTP errors (with access to the low-level error data)?

**Comment by TimothyBJacobs** [comment]
> I guess the request will just fail for these other media. I wonder if it's better to not check. Potentially, we might need post processing for other medias in the future. The check in a generic way in the middleware might be a bit complex to do. How are you checking in the media library?

For other types of media the request will succeed and return the results of a `GET` request to the item with an `edit` context.

> Also, I personally feel that instead of randomly trying to post process 5 times, the server should tell us that it's possible to post process or not using a header.

The server can't know whether it is possible to post process or not. Whether the subsizes can be generated just depends on if the server has enough resources at the moment to process the request.

> If that makes it easier/better lets add it back (just need to add the custom header in post_process_item()).

I'd like to not add any more native `header` calls to the REST API endpoints if at all possible. Looking at the patch, I don't see why it'd be technically required.

> TBD: shall we remove the create-image-subsizes action from the /post-process end-point? Looking at it again, it's not strictly necessary to have an "action" there. The client can decide whether to retry post-processing after a HTTP 500 error, and the server can "see" what kind of file was just uploaded and take appropriate action.

I'm -1 to this because it is less forward compatible. When more post processing actions are added in the future, distinguishing between them will be more complex if we didn't require a specific action to be specified up until them.

I think the `create-image-subsizes` would be fine for pdf thumbs. What types of subsizes wouldn't it make sense for? I guess resizing videos perhaps? But I think generating different video sizes would be an optional post processing step.

-----

You can also check for `missing_image_sizes` by doing a `GET` request with an `edit` context to see if progress is being made when calling `post-process`.

**Comment by azaozz** [comment]
> For other types of media the request will succeed and return the results of a GET request to the item with an edit context.

If we have a specific action, shouldn't it return an error instead saying that the action is not supported for the file type? 

Also, if eventually there are more actions in the future, what happens when a specific action is not available for a specific file type? Should be an error, right? For example setting "create-image-subsizes" on a text file upload or "resize-video" action on an image upload? WP supports uploading of many file types, only few of them would need any post-processing.

> I'm -1 to this because it is less forward compatible.

Yeah, may make sense to have an action there for the eventual case more than one type of post processing is available for a specific file type. For example for videos there may be: "optimize video" and "create video thumbnail". Then eventually the client may decide to do one type of post-processing but skip the other?

**Comment by youknowriad** [comment]
It seems to me that the endpoint has been design with the thinking that the post-processing is not really necessary so it's the responsibility of the client to know and ask for the right post-processing (action) on failures. But what happens if we don't post-process?  Is it really optional?

The thinking with my suggestions is that post-processing is not optional and the client shouldn't have to decide what post-process to perform or not, it should just tell the server post-process until it finishes.

---

I can workaround the current design (I'll refresh the patch) and I don't think changes are necessarily required for WordPress 5.3 but I wanted to clarify the thinking here.

**Comment by youknowriad** [comment]
I updated the patch,  now it handles deletions properly and also shows the error properly. 
It's still a middleware because I think that's something generic (ideally it's just one request, the fact that it's multiple is an implementation detail of the server not capable of running on a single request). I personally think it's ready.

Noting that I won't be able to work on this a lot more, since today is my last day of work before RC.

**Comment by azaozz** [comment]
Testing this again, it seems to work properly here:
retries2

For this test the timeout in PHP was set to 3 sec. Note how it timed out three times and succeeded on the forth request to `post-process`.

**Comment by azaozz** [comment]
> The thinking with my suggestions is that post-processing is not optional and the client shouldn't have to decide what post-process to perform or not, it should just tell the server post-process until it finishes.

Yes, that was the original idea too. The client should "tell" the server to finish post-processing if there was a server error. The only thing that may still be needed would be to do this only on relevant (HTTP 500) errors. For example the server would not be able to do anything if the error was with permissions, HTTP 400 or 403, etc.

> I can workaround the current design (I'll refresh the patch) and I don't think changes are necessarily required for WordPress 5.3 but I wanted to clarify the thinking here.

Right. I'm thinking it would be ideal if this is fixed in the end-point now, before releasing it in WP 5.3. The `create-image-subsizes` "action" should be removed. The server "knows" what needs to be done to post-process the upload, there is no need to limit this additionally. Then, if something like that is needed in the future, support for an optional "action" can be added. 

Alternatively the `create-image-subsizes` action can be made optional now, but I'm somewhat wary about adding something that is not needed/used.

**Comment by getsource** [comment]
Thanks so much @youknowriad @azaozz! I hadn't commented earlier because I wanted to wait until I got this to work, but realized the feedback might be helpful.

The error message looks much better, so I think this would be helpful merged.

However, I've been able to simulate failure (attachment created, but sizes not finished) by reducing the `max_execution_time`, but can't get this to auto-retry on the new WordPress Docker env. Going to try today in VVV and see if I can track it down.

**Comment by youknowriad** [comment]
Pushed a tweak to limit the behavior to 500 errors.
As I said earlier, we're going to do the Gutenberg RC today and include the fixes in WP 5.3 RC. So our deadline here is in like 4 hours.

**Comment by getsource** [comment]
As an update:

I think it might be because continue is working on 500 errors, but not on 502, which is what I’m getting on a partial Upload on the core test environments.

**Comment by getsource** [comment]
This is currently working for me and retrying / finishing images that 500. Nice!

I'm working on testing adding 502 (core patch in testing here: https://core.trac.wordpress.org/attachment/ticket/47872/47872.2.diff), but unfortunately have to stop for the night.

Thanks again for all your work on this!

**Comment by azaozz** [comment]
> I'm working on testing adding 502...

Good catch :) 502s may be caused by other problems but better to retry anyway, for cases where they are caused by PHP fatal errors on the original server. Going to add the above patch to core.

**Comment by azaozz** [comment]
With this, think this is ready. Still hoping the REST API team will reconsider removing the `create-image-subsizes` "action" (or at least making it optional) from the end-point as it doesn't serve any purpose.

File: packages/api-fetch/src/index.js [hunk 2]
```diff
@@ -80,48 +77,10 @@ const defaultFetchHandler = ( nextOptions ) => {
...
-					const unknownError = {
-						code: 'unknown_error',
-						message: __( 'An unknown error occurred.' ),
-					};
-
-					throw error || unknownError;
-				} );
-		} );
+		.catch( ( response ) => parseAndThrowError( response, parse ) )
+		.then( ( response ) => parseResponseAndNormalizeError( response, parse ) );
```
Comment (epiqueras) [inline packages/api-fetch/src/index.js]: This deserves inline comments to explain the flow.

File: packages/api-fetch/src/index.js [hunk 3]
```diff
@@ -179,5 +138,6 @@ apiFetch.createNonceMiddleware = createNonceMiddleware;
 apiFetch.createPreloadingMiddleware = createPreloadingMiddleware;
 apiFetch.createRootURLMiddleware = createRootURLMiddleware;
 apiFetch.fetchAllMiddleware = fetchAllMiddleware;
+apiFetch.mediaUploadMiddleware = mediaUploadMiddleware;
```
Comment (epiqueras) [inline packages/api-fetch/src/index.js]: Should we add this to the README middleware section?

File: packages/api-fetch/src/utils/response.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/**
```
Comment (epiqueras) [inline packages/api-fetch/src/utils/response.js]: The flow here could be simplified a lot and more JSDOC and inline comments should be added.

Let's add a note, but not block this PR so we can get it in the RC.

File: packages/api-fetch/src/middlewares/media-upload.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+				} );
+
+				return Promise.reject();
+			} );
+	};
+
+	return next( { ...options, parse: false } )
+		.catch( ( response ) => {
+			const attachmentId = response.headers.get( 'x-wp-upload-attachment-id' );
+			if ( response.status === 500 && attachmentId ) {
```
Comment (epiqueras) [inline packages/api-fetch/src/middlewares/media-upload.js]: Do we also need to test for `502`s here?

cc @azaozz

File: packages/api-fetch/src/middlewares/media-upload.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+			parse: false,
+		} )
+			.catch( () => {
+				if ( retries < maxRetries ) {
+					return postProcess( attachmentId );
+				}
+				next( {
+					path: `/wp/v2/media/${ attachmentId }?force=true`,
+					method: 'DELETE',
+				} );
```
Comment (epiqueras) [inline packages/api-fetch/src/middlewares/media-upload.js]: This could throw right?

File: packages/api-fetch/src/middlewares/media-upload.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+						return Promise.reject( {
+							code: 'post_process',
+							message: __( 'Media upload failed. If this is a photo or a large image, please scale it down and try again.' ),
+						} );
+					}
+
+					return Promise.reject( response );
+				} );
+			}
+			return Promise.reject( response );
```
Comment (epiqueras) [inline packages/api-fetch/src/middlewares/media-upload.js]: I don't think anything will parse this error.

**Review by epiqueras — APPROVED** [review]
Things seem to work in all my tests, but we should open an issue to refactor this code and make it easier to audit.

File: packages/api-fetch/src/middlewares/media-upload.js (hunk 6 above)
Comment (youknowriad) [reply packages/api-fetch/src/middlewares/media-upload.js]: I think we don't care if this throws, we need to show the post-processing error anyway in that case.

File: packages/api-fetch/src/middlewares/media-upload.js (hunk 5 above)
Comment (azaozz) [reply packages/api-fetch/src/middlewares/media-upload.js]: Yes. Based on @mikeselander's testing 502s can also mean fatal PHP errors. Going to add that to core's media modal as well.

File: packages/api-fetch/src/middlewares/media-upload.js (hunk 7 above)
Comment (youknowriad) [reply packages/api-fetch/src/middlewares/media-upload.js]: Good catch we need to wrap it with `parseAndThrowError`

**Comment by kadamwhite** [comment]
> TBD: shall we remove the create-image-subsizes action from the /post-process end-point? Looking at it again, it's not strictly necessary to have an "action" there. Both client and server "know" what file type was just uploaded at this point. The client can decide whether to retry post-processing after a HTTP 500 error, and the server would "know" what action to take for that file.
> ...
> The create-image-subsizes "action" should be removed. The server "knows" what needs to be done to post-process the upload, there is no need to limit this additionally. Then, if something like that is needed in the future, support for an optional "action" can be added.

I disagree, and continue to think making the action required is a good course.  My feeling is that removing it makes the most sense only if we were beginning to consider this endpoint more of a resource of itself, rather than an action trigger. If it were `:id/sizes` or something, I could imagine assuming that the endpoint would "know" what was needed and take appropriate secondary action if the information wasn't there; it'd be a bit non-standard but not too hard to work with, in my opinion.

However, if we remain with post-process, as I think we should, I think that the action does serve a useful purpose. It's a small thing to build a wrapper method around so I don't mind the verbosity, and it would be much easier to make this the default later than to add the need to pass an action later. For that reason I believe we should stick with the solution of passing a specific action.

> TBD: shall we pass-through the new attachment ID on requests to the /post-process end-point?

Haven't checked the patch to see if this is being included but I don't think it would need to be, as the ID should be available as a route parameter.

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad, was this change already backported to WordPress core for 5.3 release or we will need to backport in 5.4?

**Comment by youknowriad** [comment]
@jorgefilipecosta yes 👍

**Pull request merged**